pkg/logs/decoder/gdecoder.go

```diff
@@ -0,0 +1,77 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package decoder
+
+import "time"
+
+// GDecoder should replace current Decoder object.
+type GDecoder struct {
+	InputChan     chan *Input
+	OutputChan    chan *Output
+	lineGenerator LineGenerator
+}
+
+func (d *GDecoder) Start() {
+	d.lineGenerator.Start()
+}
+
+func (d *GDecoder) Stop() {
+	close(d.InputChan)
+}
+
+const defaultMaxDecodeLength = 1024 * 1024 // 1MB
+const defaultMaxSendLength = 256 * 1024    // 256KB
+
+type LineHandlerScheduler struct {
+	inputChan    chan *RichLine
+	flushTimeout time.Duration
+	lineHandler  Handler
+}
+
+func (l *LineHandlerScheduler) Stop() {
+	close(l.inputChan)
+}
+
+func (l *LineHandlerScheduler) Handle(line *RichLine) {
+	l.inputChan <- line
+}
+
+func (l *LineHandlerScheduler) Start() {
+	go l.run(l.lineHandler.Handle, l.lineHandler.SendResult)
+}
+
+func (l *LineHandlerScheduler) run(handle func(line *RichLine), sendResult func()) {
+	timer := time.NewTimer(l.flushTimeout)
+	// cleanup
+	defer func() {
+		l.lineHandler.SendResult()
+		l.lineHandler.Cleanup()
```

Comment:
I think the cleanup should be internal

Reply:
As we discussed, it can not be internal. we will need to call it from partial line handler which will be located in kubernetes package.